### PR TITLE
Cont. of Fix projection and retraction in fixed rank matrices

### PR DESCRIPTION
**Pull request by kellertuer**
Since #340 went stale by the original PR contributor, this is the followup to finish the PR.

### CODE REVIEW

**Comment by kellertuer** [comment]
While I think I got most things working, calling `project` fails to allocate a `UMTVector` currently and the allocation jungle is something, I still do not get right until now. I hope I had set `allocation_result_type` correctly? But that seems to be not enough?

Since it is now an embedded manifold (with embed and project) I also introduced `get_embedding` and added Tests for the new functions. So up to the allocation this should – hopefully – also be fine.

**Comment by kellertuer** [comment]
Found my mistake, so now I think this should be good to go.

**Comment by kellertuer** [comment]
The one additional line we miss here Is an `@view`s macro it seems; so I think this is good to check, the two things that needed to be fixed are done, all points from review in the old PR, too. From my side I think this is ready to merge.

**Review by mateuszbaran — APPROVED** [review]

**Comment by kellertuer** [comment]
@sethaxen since you had some comments on the old PR, do you think this is ok now?

File: src/Manifolds.jl [hunk 1]
```diff
@@ -1,5 +1,6 @@
 module Manifolds
 
+using LinearAlgebra: AbstractMatrix
```
Comment (mateuszbaran) [inline src/Manifolds.jl:3]:
```suggestion
```
I suspect it's something automatically inserted by VSCode?

Comment (kellertuer) [reply src/Manifolds.jl:3]: Or a new code formatter? I deleted it once but might have missed it this time. But yes, this was automatically added.

Comment (mateuszbaran) [reply src/Manifolds.jl:3]: I don't think formatter does that, see https://github.com/julia-vscode/julia-vscode/issues/1925 .

File: src/manifolds/FixedRankMatrices.jl [hunk 2]
```diff
@@ -284,19 +360,34 @@ of dimension `m`x`n` of rank `k`, namely
...
     return (m + n - k) * k * real_dimension(𝔽)
 end
 
+function Base.one(p::SVDMPoint)
+    return SVDMPoint(
+        one(zeros(size(p.U, 1), size(p.U, 1))),
+        ones(length(p.S)),
+        one(zeros(size(p.Vt, 2), size(p.Vt, 2))),
+        length(p.S),
+    )
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:375]: Why not:
```suggestion
    k = length(p.S)
    return SVDMPoint(one(p.U), one.(p.S), one(p.Vt), length(p.S))
```

File: src/manifolds/FixedRankMatrices.jl [hunk 3]
```diff
@@ -284,19 +360,34 @@ of dimension `m`x`n` of rank `k`, namely
...
+end
+function Base.one(X::UMVTVector)
+    return UMVTVector(
+        one(zeros(size(X.U, 1), size(X.U, 1))),
+        one(zeros(size(X.M))),
+        one(zeros(size(X.Vt, 2), size(X.Vt, 2))),
+        size(X.M, 1),
+    )
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:383]:
```suggestion
    return UMVTVector(one(X.U), one(X.M), one(X.Vt), size(X.M, 1))
```

**Review by sethaxen — COMMENTED** [review]
I didn't really review the math, but I had only one comment on the constructors of the custom points/vectors. Is all of the new code covered by the tests?

**Comment by kellertuer** [comment]
Thanks for taking a look, I hope I addressed all the things you asked for over in #340; concerning the coverage, I will check once the current tests ran.

File: src/manifolds/FixedRankMatrices.jl [hunk 4]
```diff
@@ -284,19 +360,25 @@ of dimension `m`x`n` of rank `k`, namely
...
+
 @doc raw"""
     project(M, p, A)
-    project(M, p, X)
 
-Project the matrix $A ∈ ℝ^{m,n}$ or a [`UMVTVector`](@ref) `X` from the embedding or
-another tangent space onto the tangent space at $p$ on the [`FixedRankMatrices`](@ref) `M`,
-further decomposing the result into $X=UMV$, i.e. a [`UMVTVector`](@ref).
+Project the matrix ``A ∈ ℝ^{m,n}`` or from the embedding the tangent space at ``p`` on the [`FixedRankMatrices`](@ref) `M`,
+further decomposing the result into ``X=UMV``, i.e. a [`UMVTVector`](@ref).
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:381]: Is this right?
```suggestion
further decomposing the result into ``X=UMV^\mathrm{H}``, i.e. a [`UMVTVector`](@ref).
```

File: src/manifolds/FixedRankMatrices.jl [hunk 5]
```diff
@@ -1,35 +1,35 @@
...
+singular values of ``p = U_p S V_p^\mathrm{H}``, where ``\cdot^{\mathrm{H}}`` denotes the complex conjugate transpose or
+Hermitian. In other words, ``U`` and ``V`` are from the manifolds [`Stiefel`](@ref)`(m,k,𝔽)` and [`Stiefel`](@ref)`(n,k,𝔽)`,
 respectively; see [`SVDMPoint`](@ref) for details.
 
-The tangent space $T_p \mathcal M$ at a point $p ∈ \mathcal M$ with $p=USV^\mathrm{H}$
+The tangent space ``T_p \mathcal M`` at a point ``p ∈ \mathcal M`` with ``p=U_p S V_p^\mathrm{H}``
 is given by
 ````math
-T_p\mathcal M = \bigl\{ UMV^\mathrm{T} + U_pV^\mathrm{H} + UV_p^\mathrm{H} :
+T_p\mathcal M = \bigl\{ U_p M V_p^\mathrm{T} + U_X V_p^\mathrm{H} + U_p V_X^\mathrm{H} :
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:20]: This too?
```suggestion
T_p\mathcal M = \bigl\{ U_p M V_p^\mathrm{H} + U_X V_p^\mathrm{H} + U_p V_X^\mathrm{H} :
```

File: src/manifolds/FixedRankMatrices.jl [hunk 6]
```diff
@@ -51,20 +51,20 @@ end
     SVDMPoint <: AbstractManifoldPoint
 
 A point on a certain manifold, where the data is stored in a svd like fashion,
-i.e. in the form $USV^\mathrm{H}$, where this structure stores $U$, $S$ and
-$V^\mathrm{H}$. The storage might also be shortened to just $k$ singular values
-and accordingly shortened $U$ (columns) and $V^\mathrm{T}$ (rows).
+i.e. in the form ``USV^\mathrm{H}``, where this structure stores ``U``, ``S`` and
+``V^\mathrm{H}``. The storage might also be shortened to just ``k`` singular values
+and accordingly shortened ``U`` (columns) and ``V^\mathrm{T}`` (rows).
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:56]:
```suggestion
and accordingly shortened ``U`` (columns) and ``V^\mathrm{H}`` (rows).
```

File: src/manifolds/FixedRankMatrices.jl [hunk 7]
```diff
@@ -247,6 +269,53 @@ function check_vector(
...
+    copyto!(X.M, Y.M)
+    copyto!(X.Vt, Y.Vt)
+    return X
+end
+
+@doc raw"""
+    embed(::FixedRankMatrices, p::SVDMPoint)
+
+Embed the point `p` from its `SVDMPoint` representation into the set of ``m×n`` matrices
+by computing ``USV^{\mathrm{T}}``.
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:289]:
```suggestion
by computing ``USV^{\mathrm{H}}``.
```

File: src/manifolds/FixedRankMatrices.jl [hunk 8]
```diff
@@ -247,6 +269,53 @@ function check_vector(
...
+
+@doc raw"""
+    embed(M::FixedRankMatrices, p, X)
+
+Embed the tangent vector `X` at point `p` in `M` from
+its [`UMVTVector`](@ref) representation  into the set of ``m×n`` matrices.
+
+The formula reads
+```math
+U_pMV_p^{\mathrm{T}} + U_XV_p^{\mathrm{T}} + U_pV_X^{mathrm{T}}
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:305]:
```suggestion
U_pMV_p^{\mathrm{H}} + U_XV_p^{\mathrm{H}} + U_pV_X^{\mathrm{H}}
```

**Comment by sethaxen** [comment]
> Thanks for taking a look, I hope I addressed all the things you asked for over in #340; concerning the coverage, I will check once the current tests ran.

No problem! Sorry for the delay. I think all of my earlier comments have been addressed.

**Comment by kellertuer** [comment]
Perfect, then I will check for coverage of the new parts as soon as the test have finished – and don't worry about the delay, the original opener of the PR did never respond back for far far longer.

**Comment by kellertuer** [comment]
The one missed line seems to stem from a `@views` line that is actually covered but not recognised.

File: src/manifolds/FixedRankMatrices.jl [hunk 9]
```diff
@@ -328,30 +406,46 @@ Compute an SVD-based retraction on the [`FixedRankMatrices`](@ref) `M` by comput
 ````math
     q = U_kS_kV_k^\mathrm{H},
 ````
-where $U_k S_k V_k^\mathrm{H}$ is the shortened singular value decomposition $USV=p+X$,
-in the sense that $S_k$ is the diagonal matrix of size $k × k$ with the $k$ largest
-singular values and $U$ and $V$ are shortened accordingly.
+where ``U_k S_k V_k^\mathrm{H}`` is the shortened singular value decomposition ``USV=p+X``,
```
Comment (sethaxen) [inline src/manifolds/FixedRankMatrices.jl:409]:
```suggestion
where ``U_k S_k V_k^\mathrm{H}`` is the shortened singular value decomposition ``USV^\mathrm{H}=p+X``,
```

**Review by sethaxen — APPROVED** [review]
I think all this needs is this docs fix and a version bump and is then ready to merge.

**Pull request merged**